samples/read_bam.c

```diff
@@ -0,0 +1,169 @@
+#include <htslib/sam.h>
+
+/// print_usage - print the demo_usage
+/** @param fp pointer to the file / terminal to which demo_usage to be dumped
+returns nothing
+*/
+static void print_usage(FILE *fp)
+{
+    fprintf(fp, "Usage: read_bam [-n position] [-i infile] [<infile>]\n\
+Shows the alignment in given position\n");
+}
+
+/// main_demo - start of the demo
+/** @param argc - count of arguments
+ *  @param argv - pointer to array of arguments
+returns 1 on failure 0 on success
+*/
+int main(int argc, char *argv[])
+{
+    static const struct option lopts[] = {
+        {NULL, 0, NULL, 0}
+    };
+
+    const char *inname = NULL;
+    int c = 0, ret = EXIT_FAILURE, idx = 0;
+    sam_hdr_t *in_samhdr = NULL;
+    samFile *infile = NULL;
+
+    int ret_r = 0, i = 0, cnt = 0;
+    bam1_t *bamdata = NULL;
+    uint8_t *data = NULL;
+    uint32_t *cigar = NULL;
+
+
+    //read_bam [-n position] [-i infile] [<infile>]
+    while ((c = getopt_long(argc, argv, "i:n:", lopts, NULL)) >= 0)
+    {
+        switch (c)
+        {
+        case 'i':                           //in file name
+            inname = optarg;
+            break;
+        case 'n':                           //position
+            idx = atoi(optarg);
+            break;
+        case '?':                           //unknown options found!
+            print_usage(stderr);
+            goto end;
+            break;
+        }
+    }
+
+    if (argc == 1 && isatty(STDIN_FILENO)) {
+        print_usage(stdout);
+        ret = EXIT_SUCCESS;
+        goto end;
+    }
+
+    if (!inname && ((argc - optind) == 1))
+        inname = argv[optind];
+    if (!inname)
+        inname = "-";
+
+    bamdata = bam_init1();                  //bam data initialization
+    if (!bamdata)
+    {
+        fprintf(stderr, "Failed to allocate data memory!\n");
+        goto end;
+    }
+
+    //open input files
+    infile = sam_open(inname, "r");
+    if (!infile) {
+        fprintf(stderr, "Could not open %s\n", inname);
+        goto end;
+    }
+    //read header
+    in_samhdr = sam_hdr_read(infile);
+    if (!in_samhdr)
+    {
+        fprintf(stderr, "Failed to read header from file!\n");
+        goto end;
+    }
+
+    while (0 <= (ret_r = sam_read1(infile, in_samhdr, bamdata)))
+    {
+        if (idx != cnt++)
+            continue;
+
+        //QNAME FLAG RNAME POS MAPQ CIGAR RNEXT PNEXT TLEN SEQ QUAL [TAG:TYPE:VALUE]…
+        fprintf(stdout, "%s", bam_get_qname(bamdata));                                  //get the query name using the macro
+        fprintf(stdout, "\t%d", bamdata->core.flag);                                    //flag is available in core structure
+        fprintf(stdout, "\t%s", sam_hdr_tid2name(in_samhdr, bamdata->core.tid));        //retrieves the target name using the value in bam and by referring the header
+        fprintf(stdout, "\t%d", bamdata->core.pos + 1);                                 //internally position is 0 based and on text output / SAM it is 1 based
```

Comment:
%d here is too small.  It's a 64-bit number usually, but we normally use `%"PRIhts_pos"` instead of `%d` which expands up to %ld or %lld as required by the system.

Reply:
Updated